### PR TITLE
Fix lifting of arguments with -coverage-out

### LINKED ISSUE #15078: Code coverage crashes at runtime
I was testing the new coverage support for scala3 with:

The latest version of `sbt-scoverage`
https://github.com/scoverage/sbt-scoverage/tree/V2

with https://github.com/scoverage/sbt-scoverage/blob/V2/src/main/scala/scoverage/ScoverageSbtPlugin.scala#L162 set to 3.2.0-RC1-bin-SNAPSHOT

This call crashes with the instrumented code:

https://github.com/MasseGuillaume/porc-epic/blob/main/src/test/scala/porcEpic/unit/VisualizationTest.scala#L53

https://github.com/MasseGuillaume/porc-epic/blob/07157829817afd03d286f056f3c0dda3fd5e6eca/src/main/scala/porcEpic/Visualization.scala#L59-L80

## Compiler version

latest (main branch)

## Minimized code

```scala
import java.nio.file._
import java.io.File

extension (v: Int) {
  def save(): File = {
    Files.createTempDirectory("porc-epic").toFile
  }
}

1.save()

```

## Output (click arrow to expand)
<details>
java.lang.ClassCastException: class scala.collection.immutable.ArraySeq$ofRef cannot be cast to class [Ljava.nio.file.attribute.FileAttribute; (scala.collection.immutable.ArraySeq$ofRef is in unnamed module of loader sbt.internal.ScalaLibraryClassLoader @53bcaacb; [Ljava.nio.file.attribute.FileAttribute; is in module java.base of loader 'bootstrap')
    at porcEpic.Visualization$package$.save(Visualization.scala:61)
    at porcEpic.unit.VisualizationTest.$init$$$anonfun$2(VisualizationTest.scala:91)
</details>

### LINKED ISSUE #15487: `-coverage-out` fails to compile formatted string interpolation
## Compiler version

3.2.0-RC1 with "-coverage-out:."

## Minimized code

```Scala
@main def main: Unit =
  println(hexMe(127))

def hexMe(i: Int): String =
  f"0x$i%016x"

```

## Output

```
[error] 5 |  f"0x$i%016x"
[error]   |  ^^^^^^^^^^^^
[error]   |  Expected statically known StringContext
[error] one error found
```

## Expectation
Code compiles

### LINKED ISSUE #15505: Erased arguments not handled in coverage
## Compiler version

3.1.2-RC1

## Minimized code

_test.scala_
```Scala
import scala.language.experimental.erasedDefinitions

erased def e(x: String): String = "x".toString
def foo(erased a: String)(b: String): String = b
def bar(x: String): String = x

def test: Unit = foo(e("aaa"))(bar("xyz"))
```

## Output

```scala
> scalac test.scala -coverage-out:. -Xprint:instrumentCoverage
[[syntax trees at end of        instrumentCoverage]] // tests/coverage/run/erased-def/test.scala
package <empty> {
  import scala.language.experimental.erasedDefinitions
  final lazy module val test$package: test$package = new test$package()
  @SourceFile("tests/coverage/run/erased-def/test.scala") final module class 
    test$package
  () extends Object() { this: test$package.type =>
    private def writeReplace(): AnyRef = 
      new scala.runtime.ModuleSerializationProxy(classOf[test$package.type])
    erased def e(x: String): String = "x".toString()
    def foo(erased a: String)(b: String): String = 
      {
        scala.runtime.coverage.Invoker.invoked(0, ".")
        b
      }
    def bar(x: String): String = 
      {
        scala.runtime.coverage.Invoker.invoked(1, ".")
        x
      }
    def test: Unit = 
      {
        scala.runtime.coverage.Invoker.invoked(5, ".")
        {
          {
            val a$1: String = 
              {
                scala.runtime.coverage.Invoker.invoked(2, ".")
                e("aaa")
              }
            val b$1: String = 
              {
                scala.runtime.coverage.Invoker.invoked(3, ".")
                bar("xyz")
              }
            {
              scala.runtime.coverage.Invoker.invoked(4, ".")
              foo(a$1)(b$1)
            }
          }
          ()
        }
      }
  }
}

-- Error: tests/coverage/run/erased-def/test.scala:7:21 ------------------------
7 |def test: Unit = foo(e("aaa"))(bar("xyz"))
  |                     ^
  |                     method e is declared as `erased`, but is in fact used
1 error found
```

## Expectation

This erased arguments should not be instrumented/lifted and therefore should not emit the error.

### PR DESCRIPTION
**Pull request by TheElectronWill**
Fixes #15078, fixes #15487, fixes #15505

Summary:
- `LiftCoverage` extended `LiftComplex`, it now extends `LiftImpure` and forces the lifting of non-erased applications
- `StringContext.s, f, raw` are now excluded from lifting
- `ElimRepeated` didn't handle lifted arguments properly, because it assumed that every argument was a `Typed` tree, which is not the case for arguments lifted by the coverage transformation. This is now fixed.

Most of the added lines are tests :)

## How I fixed it

Let's see if I can imitate the awesome explanations of Martin :smile:

`InstrumentCoverage` modifies the AST just after the pickling, and insert calls to `scala.runtime.coverage.Invoker.invoked` that record, at runtime, which methods are called.
See an example [here](https://github.com/lampepfl/dotty/issues/15078#issuecomment-1127592307). To lift functions' arguments, it uses `LiftCoverage`, which lives with its `LiftXX` siblings in `EtaExpansion.scala`.

`LiftCoverage` previously extended `LiftComplex`. A quick look at the definition of `LiftComplex` tells me that it lifts everything, except trees that are constants or have the maximal purity `PurePath`. And [`TreeInfo.exprPurity`](https://github.com/lampepfl/dotty/blob/fb6d00467dab102a5f6732ade4900da24940d86b/compiler/src/dotty/tools/dotc/ast/TreeInfo.scala#L418) says that erased things are `Pure`, not `PurePath`! It seems that `LiftImpure` would be more appropriate than `LiftComplex`. Therefore, I happily type `LiftCoverage extends LiftImpure`. But stopping there would introduce another bug!

For example, given this code:
```scala
def f() = throw RuntimeException() // throws an exception
def g() = 0 // pure function
someFunction(f(), g())
```
We want to generate something like this, even if g has no side effects:
```scala
Invoker.invoked("f", ...)
val _f = f()
Invoker.invoked("g", ...)
val _g = g()
Invoker.invoked("someFunction", ...)
someFunction(_f, _g)
```
This way, f will throw its exception and prevent g from being marked as covered, which would be wrong.

Therefore, all non-erased applications must be lifted. This means that, when lifting arguments, `LiftCoverage.noLift` must return `false` for all non-erased `Apply`:
```scala
override def noLift(expr: tpd.Tree)(using Context) =
  if liftingArgs then noLiftArg(expr) else super.noLift(expr)

private def noLiftArg(arg: tpd.Tree)(using Context): Boolean =
  arg match
    case a: tpd.Apply => a.symbol.is(Erased) // don't lift erased applications, but lift all others
    // the other cases recurse on more complex trees that may contain applications
```

Similarly, string interpolators `s`, `f` and `raw` shouldn't be lifted. I've excluded those three functions from lifting in `InstrumentCoverage.needsLift`.

Finally, I took a look at the mysterious issue #15078.
For a call to this java method:
```scala
Path createTempDirectory(String prefix, FileAttribute<?>... attrs)
```
`-Ycheck-all` reported the following problem:
```scala
java.lang.AssertionError: assertion failed: Found:    ($2$ : java.nio.file.attribute.FileAttribute[?]*)
Required: Array[? <: java.nio.file.attribute.FileAttribute[?]]
```
And the code crashed at runtime...

Yet again, it only happened when the arguments were lifted by `-coverage-out`. A quick minimization revealed that the problem was the handling of the java varargs, in the presence of lifted arguments.

I know that `ElimRepeated` is the phase that takes care of handling varargs. Some debugging showed that [`ElimRepeated.transformApply`](https://github.com/lampepfl/dotty/blob/fb6d00467dab102a5f6732ade4900da24940d86b/compiler/src/dotty/tools/dotc/transform/ElimRepeated.scala#L126) was triggered, with:
- `isWildcardStarArg(arg)`: true
- `arg` of type `Ident`

`isWildcardStarArg` properly handled `Ident`, but not `ElimRepeated`, which assumed that everything was a `Typed` tree. Fixing that made the bug vanish.

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala [hunk 1]
```diff
@@ -273,24 +273,34 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
      * should not be changed to {val $x = f(); T($x)}(1) but to {val $x = f(); val $y = 1; T($x)($y)}
      */
     private def needsLift(tree: Apply)(using Context): Boolean =
-      def isBooleanOperator(fun: Tree) =
-        // We don't want to lift a || getB(), to avoid calling getB if a is true.
-        // Same idea with a && getB(): if a is false, getB shouldn't be called.
-        val sym = fun.symbol
-        sym.exists &&
+      inline def isShortCircuitedOp(sym: Symbol) =
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala:276]: The JVM should already do a good job at inlining short methods, so I don't think `inline` is useful here or in the next method.

File: compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala [hunk 2]
```diff
@@ -273,24 +273,34 @@ class InstrumentCoverage extends MacroTransform with IdentityDenotTransformer:
...
         sym == defn.Boolean_&& || sym == defn.Boolean_||
 
-      def isContextual(fun: Apply): Boolean =
-        val args = fun.args
-        args.nonEmpty && args.head.symbol.isAllOf(Given | Implicit)
+      inline def isCompilerIntrinsic(sym: Symbol) =
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala:279]: This definition could be moved to the companion object of StringInterpolatorOpt, and used in that phase at https://github.com/lampepfl/dotty/blob/b0671c030bbf9803b5f340b9cc731e48a9b1a5fd/compiler/src/dotty/tools/dotc/transform/localopt/StringInterpolatorOpt.scala#L34 for example.

**Review by smarter — CHANGES_REQUESTED** [review]
Otherwise LGTM!

File: compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala (hunk 2 above)
Comment (TheElectronWill) [reply compiler/src/dotty/tools/dotc/transform/InstrumentCoverage.scala]: done :)

**Comment by smarter** [comment]
Looks like the ci is failing with a coverage test check file mismatch

**Comment by TheElectronWill** [comment]
Ah yes, it must be because of #15504. I'll update the checkfiles

**Review by smarter — APPROVED** [review]
LGTM! If you have the time, I think this is something that would be worth backporting to the release-3.2.0 branch too (and https://github.com/lampepfl/dotty/pull/15504 too presumably?)

**Comment by TheElectronWill** [comment]
> LGTM! If you have the time, I think this is something that would be worth backporting to the release-3.2.0 branch too (and #15504 too presumably?)

Sure! Here you are: #15573 :)

**Pull request merged**